scenarios/kind/node_scenarios_example.yml

```diff
@@ -0,0 +1,16 @@
+node_scenarios:
+  - actions:                                                        # node chaos scenarios to be injected
+    - node_stop_start_scenario
+    node_name: kind-worker                                          # node on which scenario has to be injected; can set multiple names separated by comma
+    # label_selector: node-role.kubernetes.io/worker                # when node_name is not specified, a node with matching label_selector is selected for node chaos scenario injection
+    instance_count: 1                                               # Number of nodes to perform action/select that match the label selector
+    runs: 1                                                         # number of times to inject each scenario under actions (will perform on same node each time)
+    timeout: 120                                                    # duration to wait for completion of node scenario injection
+    cloud_type: docker                                                # cloud type on which Kubernetes/OpenShift runs
+  - actions:
+    - node_reboot_scenario
+    node_name: kind-worker
+    # label_selector: node-role.kubernetes.io/infra
```

Comment:
I'm getting the below error when running the node reboot scenario (no issues with start stop)

```
2022-12-06 12:41:00,960 [INFO] Starting node_reboot_scenario injection
2022-12-06 12:41:00,970 [INFO] Rebooting the node kind-worker2 with container ID: 7c3853a1abeb292bcb324e7a484b6c30a9f463828cc52053e032fde2eff8a75c 
2022-12-06 12:41:03,837 [INFO] Status of node kind-worker2: True
2022-12-06 12:41:53,842 [ERROR] Failed to reboot node instance. Encountered following exception: ("Connection broken: InvalidChunkLength(got length b'', 0 bytes read)", InvalidChunkLength(got length b'', 0 bytes read)). Test Failed
2022-12-06 12:41:53,842 [ERROR] node_reboot_scenario injection failed!
```

Reply:
Just to note I did use a 2 worker node kind config found [here](https://github.com/redhat-chaos/actions/blob/main/.kind-config.yml)

Reply:
I have tried the node reboot scenario with no errors.
Config file:
```
node_scenarios:
  - actions:                                                        # node chaos scenarios to be injected
    - node_reboot_scenario
    node_name: kind-worker
    # label_selector: node-role.kubernetes.io/infra
    instance_count: 1
    timeout: 120
    cloud_type: docker
```

```
% python3 run_kraken.py --config config/config_kind.yaml
:0: UserWarning: You do not have a working installation of the service_identity module: 'No module named 'service_identity''.  Please install it from <https://pypi.python.org/pypi/service_identity> and make sure all of its dependencies are satisfied.  Without the service_identity module, Twisted can perform only rudimentary TLS client hostname verification.  Many valid certificate/hostname mappings may be rejected.
 _              _              
| | ___ __ __ _| | _____ _ __  
| |/ / '__/ _` | |/ / _ \ '_ \ 
|   <| | | (_| |   <  __/ | | |
|_|\_\_|  \__,_|_|\_\___|_| |_|
                               

2022-12-07 10:54:26,228 [INFO] Starting kraken
2022-12-07 10:54:26,235 [INFO] Initializing client to talk to the Kubernetes cluster
2022-12-07 10:54:26,413 [INFO] Publishing kraken status at http://0.0.0.0:8081
2022-12-07 10:54:26,413 [INFO] Publishing kraken status at http://0.0.0.0:8081
2022-12-07 10:54:26,416 [INFO] Starting http server at http://0.0.0.0:8081

2022-12-07 10:54:26,416 [INFO] Fetching cluster info
2022-12-07 10:54:26,427 [INFO] Cluster version CRD not detected, skipping
2022-12-07 10:54:26,427 [INFO] Server URL: https://127.0.0.1:62456
2022-12-07 10:54:26,427 [INFO] Generated a uuid for the run: d1d138c8-5899-4dd7-91b2-b25e858acaf5
2022-12-07 10:54:26,427 [INFO] Daemon mode not enabled, will run through 1 iterations

2022-12-07 10:54:26,427 [INFO] Executing scenarios for iteration 0
2022-12-07 10:54:26,427 [INFO] connection set up
127.0.0.1 - - [07/Dec/2022 10:54:26] "GET / HTTP/1.1" 200 -
2022-12-07 10:54:26,428 [INFO] response RUN
2022-12-07 10:54:26,428 [INFO] Running node scenarios
2022-12-07 10:54:26,484 [INFO] Starting node_reboot_scenario injection
2022-12-07 10:54:26,493 [INFO] Rebooting the node kind-worker with container ID: eaba5a798025cf82e41f2d742cddde40e6e5ce9bac4838cbe1674ec9e04512fd 
2022-12-07 10:54:32,990 [INFO] Status of node kind-worker: True
 2022-12-07 10:58:32,591 [INFO] Node with container ID: eaba5a798025cf82e41f2d742cddde40e6e5ce9bac4838cbe1674ec9e04512fd has been rebooted
2022-12-07 10:58:32,592 [INFO] node_reboot_scenario has been successfuly injected!
2022-12-07 10:58:32,592 [INFO] Waiting for the specified duration: 60
2022-12-07 10:59:32,593 [INFO] 
2022-12-07 10:59:32,598 [INFO] 
2022-12-07 10:59:32,600 [INFO] Successfully finished running Kraken. UUID for the run: d1d138c8-5899-4dd7-91b2-b25e858acaf5. Report generated at /Users/jlema/krkn/kraken.report. Exiting
```

Container got restarted (see status column):
```
% docker ps
CONTAINER ID   IMAGE                  COMMAND                  CREATED          STATUS          PORTS                       NAMES
ab86d77e0999   kindest/node:v1.23.4   "/usr/local/bin/entr…"   16 minutes ago   Up 15 minutes   127.0.0.1:62456->6443/tcp   kind-control-plane
066859fda795   kindest/node:v1.23.4   "/usr/local/bin/entr…"   16 minutes ago   Up 15 minutes                               kind-worker2
eaba5a798025   kindest/node:v1.23.4   "/usr/local/bin/entr…"   16 minutes ago   Up 6 minutes                                kind-worker
```

The only thing that I can think of is that the restart scenario may be so fast that Kubernetes does not mark the node as not ready.

Reply:
Is there maybe a way to add in a check to see if the status time has updated since running the tests to get around this?

Reply:
The docker.* methods (start, stop, restart) are synchronous, different from AWS or baremetal actions via IPMI.
So either they finish correctly or they through a corresponding error message.
Imho there is no value added by a wait action in this synchronous scenario.